keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+YAML or JSON of the apply configuration. This has a couple important
+deficiencies:
+
+- It is a gap completeness of the type client, which provides typesafe APIs for
+  all other major methods.
+- It makes it to easy for developers to make a major, but non-obvious mistake:
+  Use the existing go structs to construct an apply configuration, serialize
+  it to JSON, and pass it to `Patch`. This can cause zero valued required
+  fields being accientally included in the apply configuration resulting
+  in fields being accidentally set to incorrect values and/or fields accidentally
+  being accidentally being clamed as owned.
+
+Both sig-api-machinery and wg-api-expression agree that this enhancement is
+required before server side apply to be promoted to GA.
+
+### Goals
+
+Introduce a typesafe, programmatic way to call server side apply using the typed
+client in client-go.
+
+Express all apply configurations in Go that can be expressed in
+YAML. Specifically, an apply request must include only field that are set by
+applier and exclude those not set by applier.
+
+Validate this enhancement meets the needs of developers:
+
+- An developer not directly involved in this enhancement successfully converts
+  a 1st party controller (one in github.com/kubernetes/kubernetes) to us this
+  enhancement.
+- A representative group of the developer community is made aware of this
+  proposed enhancement, is given early access to it via a fork of
+  controller-tools with the requisite generators, and is given the opportunity
+  to try it out and provide feedback.
+
+### Non-Goals
+
+Enhancements to client-go's dynamic client. The client-go dynamic client already
+supports Apply via Patch, which is adequate for the dynamic client.
+
+Protobuf support. Apply does not support protobuf, and it will not be added with
+this enhancement.
+
+## Proposal
+
+`Apply` functions should be included in the typed clients generated for
+client-go and should accept the apply configuration using a strongly typed
+representation, which will need to be generated for this purpose.
+
+### Risks and Mitigations
+
+#### Poor adoption
+
+Risk: Developers adoption is poor, either because the asthetics/ergonomics are
+not to their liking or the functinality is insufficient to do what they need to
+do with it. This could lead to (a) poor server side apply adoption, and/or (b)
+developers building alternate solutions.
+
+Mitigation: We are working with the kubebuilder community to
+get hands on feedback from developers to guide our design decisions around
+asthetics/egronomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)
```

Comment:
super unfortunate to have fieldManager separate from apply options, any way we can combine / embed?

Reply:
Last parameter is unnamed?

Reply:
Maybe make it "otherOptions"?

Also what if you change the order to put the deployment object last, I feel like that will help readability if someone is going to inline the object definition?

Reply:
I think it makes sense to have the fieldManager in the signature. Feels more explicit while things in options are optional.

Reply:
Fixed the last parameter. It is named, and there is also a `subresources ...string` parameter that I had accidentally omitted here. The signature is the same as `Patch` but with the `fieldManager` param required.

I'll move the fieldManager param to before the object. I'm hesitant to change the order of the params too much because there are parameter order conventions the method functions. For context:

```go
Create(ctx context.Context, deployment *v1.Deployment, opts metav1.CreateOptions) (*v1.Deployment, error)
Update(ctx context.Context, deployment *v1.Deployment, opts metav1.UpdateOptions) (*v1.Deployment, error)
UpdateStatus(ctx context.Context, deployment *v1.Deployment, opts metav1.UpdateOptions) (*v1.Deployment, error)
Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
DeleteCollection(ctx context.Context, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error
Get(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Deployment, error)
List(ctx context.Context, opts metav1.ListOptions) (*v1.DeploymentList, error)
Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error)
Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *v1.Deployment,  err error)
Apply(ctx context.Context, deployment *appsv1.DeploymentApplyConfiguration, fieldManager string, opts metav1.ApplyOptions, subresources ...string) (result *v1.Deployment, err error)
GetScale(ctx context.Context, deploymentName string, options metav1.GetOptions) (*autoscalingv1.Scale, error)
UpdateScale(ctx context.Context, deploymentName string, scale *autoscalingv1.Scale, opts metav1.UpdateOptions) (*autoscalingv1.Scale, error)
```

Reply:
I've added an `UNRESOLVED` to the KEP for how we organize the params.